src/types.c

```diff
@@ -69,7 +69,7 @@ SEXP testMsgR(SEXP status, SEXP x, SEXP k) {
     }
   }
 
-  #pragma omp parallel for if (nx*nk>1) schedule(auto) collapse(2) num_threads(getDTthreads())
+  #pragma omp parallel for schedule(auto) collapse(2) num_threads(getDTthreads(nx*nk, true))
```

Comment:
here `throttle=false`, these are meant to be used for embarrassingly parallel iterations. Also `schedule(dynamic)` should be here.

Reply:
My thinking here was for say 10 columns of 2 rows.  For that, single thread (i.e. apply throttle). For very wide (e.g. 1,000,000 columns x 2 rows) and very tall (e.g. 2 columns of 1,000,000 rows) I'm not sure how the `collapse(2)` interacts.  Needs testing.  But because it's a throttle on `nk*nk` (not just `nk`), I was thinking that mult-threading would be utilized even for 2 columns, as soon as the columns got tall enough.

Reply:
`schedule(dynamic)` sounds good. When there's a `collapse(2)` like here do you know if it's as if there one loop of `nx*nk` items to which the dynamic applies,  or is is just on the outer loop through `nx`? I'd hope the former, and that'd be really nice, but just wondered if we knew that for sure or if maybe that varies by OpenMP implementation.

Reply:
This particular place in `types.c` is just for testing messages that are used in `frollR.c`, so that loop definition should match to what is `frollR.c`. I will benchmark those cases you described.

Reply:
Regarding collapse, I think loop is collapsed into single loop, and then all other settings apply.

Benchmark on frollmean, where we use nx*nk and collapse(2). Tall data scaled up to 1e8.

Conclusions:

`getDTthreads(nx*nk, true)`:
- parallelism in wide data does not give speed up or an overhead, probably looping through 1e6 columns is still good using 40 threads rather than 1, so the overhead of omp is neglected.
- disables parallelism for tall data because nx=2 and nk=2 so 4 < 1024

`getDTthreads(nx*nk, false)`:
- for wide data same as `throttle=true`
- for tall data it uses threads very efficiently: nx=2 nk=1 (1.287s down to 0.626s); nx=2 nk=2 (2.196s down to 0.694s).

Note that throttle=true will not take `nrow` into account at all, because `nx` is columns and `nk` are window widths.
It make sense to use `throttle=false` then. The problem will still be there when doing `frollmean`/`nafill` by many groups. To address that properly we would need to have more arguments to getDTthreads, which make sense to leave for a follow up PR.

----

Script and timings below.

```r
library(data.table)
# wide (e.g. 1,000,000 columns x 2 rows)
x = setDT(lapply(1:1e6, function(x) rnorm(2L)))
#options(datatable.verbose=TRUE)

setDTthreads(1L);  a=frollmean(x, n=2L)  ## nk=1
setDTthreads(40L); a=frollmean(x, n=2L)
setDTthreads(1L);  a=frollmean(x, n=2:3) ## nk=2
setDTthreads(40L); a=frollmean(x, n=2:3)

# tall (e.g. 2 columns of 100,000,000 rows) 
x = setDT(list(rnorm(1e8L), rnorm(1e8L)))

setDTthreads(1L);  a=frollmean(x, n=20L)
setDTthreads(40L); a=frollmean(x, n=20L)
setDTthreads(1L);  a=frollmean(x, n=c(20L,30L))
setDTthreads(40L); a=frollmean(x, n=c(20L,30L))
```
```
### schedule(dynamic) getDTthreads(nx*nk, true)

## wide
setDTthreads(1L);  a=frollmean(x, n=2L)  ## nk=1
#frollfunR: processing of 1000000 column(s) and 1 window(s) took 11.747s
setDTthreads(40L); a=frollmean(x, n=2L)
#frollfunR: processing of 1000000 column(s) and 1 window(s) took 11.827s
setDTthreads(1L);  a=frollmean(x, n=2:3) ## nk=2
#frollfunR: processing of 1000000 column(s) and 2 window(s) took 23.035s
setDTthreads(40L); a=frollmean(x, n=2:3)
#frollfunR: processing of 1000000 column(s) and 2 window(s) took 23.016s

## tall
setDTthreads(1L);  a=frollmean(x, n=20L)
#frollfunR: processing of 2 column(s) and 1 window(s) took 1.502s
setDTthreads(40L); a=frollmean(x, n=20L)
#frollfunR: processing of 2 column(s) and 1 window(s) took 1.209s
setDTthreads(1L);  a=frollmean(x, n=c(20L,30L))
#frollfunR: processing of 2 column(s) and 2 window(s) took 2.291s
setDTthreads(40L); a=frollmean(x, n=c(20L,30L))
#frollfunR: processing of 2 column(s) and 2 window(s) took 2.358s

### schedule(dynamic) getDTthreads(nx*nk, false)

## wide
setDTthreads(1L);  a=frollmean(x, n=2L)  ## nk=1
#frollfunR: processing of 1000000 column(s) and 1 window(s) took 11.651s
setDTthreads(40L); a=frollmean(x, n=2L)
#frollfunR: processing of 1000000 column(s) and 1 window(s) took 11.441s
setDTthreads(1L);  a=frollmean(x, n=2:3) ## nk=2
#frollfunR: processing of 1000000 column(s) and 2 window(s) took 21.929s
setDTthreads(40L); a=frollmean(x, n=2:3)
#frollfunR: processing of 1000000 column(s) and 2 window(s) took 22.186s

## tall
setDTthreads(1L);  a=frollmean(x, n=20L)
#frollfunR: processing of 2 column(s) and 1 window(s) took 1.287s
setDTthreads(40L); a=frollmean(x, n=20L)
#frollfunR: processing of 2 column(s) and 1 window(s) took 0.626s
setDTthreads(1L);  a=frollmean(x, n=c(20L,30L))
#frollfunR: processing of 2 column(s) and 2 window(s) took 2.196s
setDTthreads(40L); a=frollmean(x, n=c(20L,30L))
#frollfunR: processing of 2 column(s) and 2 window(s) took 0.694s
```